fairlearn/postprocessing/_cvxpy_threshold_optimizer.py

```diff
@@ -0,0 +1,395 @@
+
+                # Bottom-left (0,0) and top-right (1,1) points must be kept
+                indices_to_keep[-1] = len(fpr) - 1
+                roc_curve_data = (fpr[indices_to_keep], tpr[indices_to_keep], thrs[indices_to_keep])
+
+            self._all_roc_data[g] = roc_curve_data
+
+        # Compute convex hull of each ROC curve
+        self._all_roc_hulls = dict()
+        for g in unique_groups:
+            group_fpr, group_tpr, _group_thresholds = self._all_roc_data[g]
+
+            curr_roc_points = np.stack((group_fpr, group_tpr), axis=1)
+            curr_roc_points = np.vstack((curr_roc_points, [1, 0]))  # Add point (1, 0) to ROC curve
+
+            self._all_roc_hulls[g] = roc_convex_hull(curr_roc_points)
+
+        # Find the group-wise optima that fulfill the fairness criteria
+        self._groupwise_roc_points, self._global_roc_point = compute_equal_odds_optimum(
+            groupwise_roc_hulls=self._all_roc_hulls,
+            fairness_tolerance=self.tolerance,
+            group_sizes_label_pos=group_sizes_label_pos,
+            group_sizes_label_neg=group_sizes_label_neg,
+            global_prevalence=self._global_prevalence,
+            false_positive_cost=self.false_pos_cost,
+            false_negative_cost=self.false_neg_cost,
+        )
+
+        # Construct each group-specific classifier
+        all_rand_clfs = {
+            g: RandomizedClassifier.construct_at_target_ROC(
+                predictor=self.estimator,
+                roc_curve_data=self._all_roc_data[g],
+                target_roc_point=self._groupwise_roc_points[g],
+                seed=self.seed,
+            )
+            for g in unique_groups
+        }
+
+        # Construct the global classifier (can be used for all groups)
+        self._realized_classifier = EnsembleGroupwiseClassifiers(group_to_clf=all_rand_clfs)
+        return self
+    
+    def _check_fit_status(self, raise_error: bool = True) -> bool:
+        """Checks whether this classifier has been fit on some data.
+        
+        Parameters
+        ----------
+        raise_error : bool, optional
+            Whether to raise an error if the classifier is uninitialized 
+            (otherwise will just return False), by default True.
+
+        Returns
+        -------
+        is_fit : bool
+            Whether the classifier was already fit on some data.
+
+        Raises
+        ------
+        RuntimeError
+            If `raise_error==True`, raises an error if the classifier is
+            uninitialized.
+        """
+        if self._realized_classifier is None:
+            if not raise_error:
+                return False
+
+            raise RuntimeError(
+                "This classifier has not yet been fitted to any data. "
+                "Call clf.fit(...) before this method.")
+
+        return True
+
+    def predict(self, X: np.ndarray, group: np.ndarray) -> np.ndarray:
```

Comment:
I wrote a little about data types in the issue. This is just a reminder.

Reply:
Fixed as of commit 7510e560e67bbedf25d2899f3a01ed63913aa15d